libdnf/rpm/package.cpp

```diff
@@ -313,7 +314,23 @@ std::string Package::get_repo_id() const {
 }
 
 libdnf::transaction::TransactionItemReason Package::get_reason() const {
-    return base->get_system_state().get_package_reason(get_na());
+    auto reason = base->get_system_state().get_package_reason(get_na());
```

Comment:
I prefer to add a test whether package with NA is installed first. I think it will provide a better behavior.

Reply:
I am not sure how you mean it exactly, did you mean it like this?
```c++
rpm::PackageQuery installed_query(base, rpm::PackageQuery::ExcludeFlags::IGNORE_EXCLUDES);
installed_query.filter_installed();
installed_query.filter_name({get_name()});
installed_query.filter_arch({get_arch()});
if (!installed_query.empty()) {
    auto reason = base->get_system_state().get_package_reason(get_na());

    if (reason == libdnf::transaction::TransactionItemReason::NONE) {
        return libdnf::transaction::TransactionItemReason::EXTERNAL_USER;
    }

    return reason;
}

return libdnf::transaction::TransactionItemReason::NONE;
```

I don't agree it's better, for one, it is quite a bit more expensive to always do the query, instead of only when the reason is `NONE` (a small minority of cases). Of course, in this case you don't have to search for the reason if the NA is not installed, but the assumption here is that most of the time you're querying for a NA that is installed.

Second, it changes behavior, I've created a test case for this: https://github.com/rpm-software-management/ci-dnf-stack/pull/1128/commits/a3a115cbbac414847f7ee3d04d823f679bf341da

Using the above block will break the test case, the `filesystem.x86_64` will have reason `Dependency` after the second install.

Reply:
> Using the above block will break the test case, the filesystem.x86_64 will have reason Dependency after the second install.

Do you know why it returns dep?

Reply:
> Do you know why it returns dep?

Yes, because with your approach it checks whether the NA is installed first, and it's not. There's reason `USER` in system state, but we check for the NA being installed and since it's not, we return `NONE`.

Reply:
I am sorry I overlooked something in the test. DEP is correct for `filesystem.x86_64` because the test requests to install only `basesystem` and `filesystem` was added to transaction as a dependency.

Reply:
Ok. Do you then want me to use the piece of code mentioned in https://github.com/rpm-software-management/libdnf/pull/1543#discussion_r933517720?

Reply:
Yes, if you can use it.